syntax = "proto3";

package dps;

service BrokerServer
{
  // replication of messages 
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse) {}

  // leader election for topics 
  rpc StartElection(StartElectionRequest) returns (StartElectionResponse) {}
  rpc RequestVote (RequestVoteRequest) returns (RequestVoteResponse) {}
  
  // data streaming APIs 
  rpc SendMessage(SendMessageRequest) returns (SendMessageResponse) {}
  rpc ReadMessageStream(ReadMessageRequest) returns (stream ReadMessageResponse) {}
}

message AppendEntriesRequest { }
message AppendEntriesResponse { }

message StartElectionRequest { }
message StartElectionResponse { }
message RequestVoteRequest { }
message RequestVoteResponse { }

message SendMessageRequest { }
message SendMessageResponse { }
message ReadMessageRequest { }
message ReadMessageResponse { }

service GuruServer
{
  // pub/sub clients to get broker for operations
  rpc GetBrokerForWrite(GetBrokerRequest) returns (GetBrokerResponse) {}
  rpc GetBrokerForRead(GetBrokerRequest) returns (GetBrokerResponse) {}

  rpc Subscribe(SubscribeRequest) returns (SubscribeResponse) {}

  rpc AddTopic(AddTopicRequest) returns (AddTopicResponse) {}

  // Communication with brokers
  rpc SetLeader(SetLeaderRequest) returns (SetLeaderResponse) {}
  rpc SendHeartbeat(HeartbeatRequest) returns (HeartbeatResponse) {}
}

message GetBrokerRequest { }
message GetBrokerResponse { }

message SubscribeRequest { }
message SubscribeResponse { }

message AddTopicRequest { }
message AddTopicResponse { }

message SetLeaderRequest { }
message SetLeaderResponse { }
message HeartbeatRequest {
  uint32 serverid = 1;
  uint32 clusterid = 2;
}
message HeartbeatResponse {
  bool success = 1;
}